### PR TITLE
Add an error ANOVA functions when control assays are present

### PR DESCRIPTION
**Pull request by dtopouza**
# Title: Add an error ANOVA functions when control assays are present
**Problem:** ANOVA function crashes when sum of squares equals 0. Need to have a trycatch to remind customers to exclude internal control assays and external controls from statistical analyses.

**Solution:** Added a conditional statement in olink_anova() and olink_anova_posthoc() that throws an informative error if control assays have not been removed. Added unit tests.

## Checklist
- [x] Make sure you are requesting to **pull a feature/bugfix branch** (right side). This should not be main or develop.
- [x] Make sure you are make a pull request against either **main or develop** (left side). (Requesting to main should be reserved for bugfixes and new releases)
- [x] Add or update unit tests (if applicable)
- [x] Check your code with any unit tests (Run devtools::check() locally)
- [ ] Add neccessary documentation (if applicable)

## Type of changes

What type of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue) (https://olink.atlassian.net/browse/OA-2272)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code style update (formatting, renaming)
- [ ] Documentation Update 
- [ ] Other (explain)

### CODE REVIEW

File: OlinkAnalyze/R/Olink_anova.R [hunk 1]
```diff
@@ -106,8 +106,19 @@ olink_anova <- function(df,
...
 
-  withCallingHandlers({
+  # Stop if control assays have not been removed
+  ctrl_types <- c("amp_ctrl", "ext_ctrl", "inc_ctrl")
 
+  if ("AssayType" %in% names(df)) {
+    if (any(ctrl_types %in% df$AssayType)) {
+      stop('Control assays have not been removed from the dataset.\n  Assays with AssayType != "assay" should be excluded.')
+    }
+  } else if (any(str_detect(df$Assay, "control"))) {
```
Comment (kathy-nevola) [inline OlinkAnalyze/R/Olink_anova.R:116]:
```suggestion
  } else if (any(str_detect(df$Assay, regex("control|ctrl", ignore_case = TRUE)))) {
```

File: OlinkAnalyze/R/Olink_anova.R [hunk 2]
```diff
@@ -443,9 +454,19 @@ olink_anova_posthoc <- function(df,
     stop("All effect terms must be included in the variable argument or model formula.")
   }
 
+  # Stop if control assays have not been removed
+  ctrl_types <- c("amp_ctrl", "ext_ctrl", "inc_ctrl")
```
Comment (kathy-nevola) [inline OlinkAnalyze/R/Olink_anova.R:458]: could this be changed to if any(df$AssayType != "assay")

**Review by kathy-nevola — APPROVED** [review]
Approved pending minor changes. Seems to work as is but there may be a few possible edge cases

File: OlinkAnalyze/R/Olink_anova.R [hunk 3]
```diff
@@ -106,8 +106,19 @@ olink_anova <- function(df,
     stop('The df and variable arguments need to be specified.')
   }
 
-  withCallingHandlers({
+  # Stop if control assays have not been removed
+  ctrl_types <- c("amp_ctrl", "ext_ctrl", "inc_ctrl")
 
+  if ("AssayType" %in% names(df)) {
+    if (any(ctrl_types %in% df$AssayType)) {
```
[the remark this answers is not in the document]
Comment (dtopouza) [reply OlinkAnalyze/R/Olink_anova.R:113]: Will change in next version

File: OlinkAnalyze/R/Olink_anova.R (hunk 2 above)
Comment (dtopouza) [reply OlinkAnalyze/R/Olink_anova.R:458]: Will change in next version

File: OlinkAnalyze/R/Olink_anova.R [hunk 4]
```diff
@@ -443,9 +454,19 @@ olink_anova_posthoc <- function(df,
...
+      stop('Control assays have not been removed from the dataset.\n  Assays with AssayType != "assay" should be excluded.')
+    }
+  } else if (any(str_detect(df$Assay, "control"))) {
+    stop('Control assays have not been removed from the dataset.\n  Assays with "control" in their Assay field should be excluded.')
+  }
```
[the remark this answers is not in the document]
Comment (dtopouza) [reply OlinkAnalyze/R/Olink_anova.R:466]: That's a good idea, I will add the IDs to the message.

File: OlinkAnalyze/R/Olink_anova.R [hunk 5]
```diff
@@ -443,9 +460,25 @@ olink_anova_posthoc <- function(df,
...
+  # Stop if control assays have not been removed
+  if ("AssayType" %in% names(df)) {
+    if (any(df$AssayType != "assay")) {
+      ctrl_assays <- df |>
+        dplyr::filter(AssayType != "assay")
+      
+      stop(paste0('Control assays have not been removed from the dataset.\n  Assays with AssayType != "assay" should be excluded.\n  The following control assays were found:\n  ',
+                  paste(strwrap(toString(unique(ctrl_assays$Assay)), width = 80), collapse = "\n")))
+    }
+  } else if (any(str_detect(df$Assay, regex("control|ctrl", ignore_case = TRUE)))) {
```
Comment (kathy-nevola) [inline OlinkAnalyze/R/Olink_anova.R:472]: package name needs to be specified stringr::str_detect and stringr::regex()

File: OlinkAnalyze/tests/testthat/test-Olink_anova.R [hunk 6]
```diff
@@ -5,6 +5,34 @@ load(refRes_file)
 #Load data with hidden/excluded assays (all NPX=NA)
 load(file = '../data/npx_data_format221010.RData')
 
+# Remove assays with NPX == NA from npx_data_format221010 for testing
+npx_data_format221010_no_NA <- npx_data_format221010 |>
+  dplyr::filter(!is.na(NPX))
+
+# Add dummy Extension control assays
+extension_control <- npx_data_format221010_no_NA |>
+  dplyr::filter(str_detect(Assay, "Incubation control")) |>
```
Comment (kathy-nevola) [inline OlinkAnalyze/tests/testthat/test-Olink_anova.R:14]: stringr::str_detect

File: OlinkAnalyze/tests/testthat/test-Olink_anova.R [hunk 7]
```diff
@@ -5,6 +5,34 @@ load(refRes_file)
...
+  dplyr::mutate(UniProt = gsub("INC", "EXT", UniProt))
+
+npx_data_format221010_ext_ctrl <- rbind(
+  npx_data_format221010_no_NA,
+  extension_control
+)
+
+# Add AssayType
+npx_data_format221010_AssayType <- npx_data_format221010_ext_ctrl |>
+  mutate(AssayType = case_when(
```
Comment (kathy-nevola) [inline OlinkAnalyze/tests/testthat/test-Olink_anova.R:25]: dplyr::mutate

Comment (kathy-nevola) [reply OlinkAnalyze/tests/testthat/test-Olink_anova.R:25]: and dplyr::case_when

File: OlinkAnalyze/tests/testthat/test-Olink_anova.R [hunk 8]
```diff
@@ -5,6 +5,34 @@ load(refRes_file)
...
+    str_detect(Assay, "Incubation control") ~ "inc_ctrl",
```
Comment (kathy-nevola) [inline OlinkAnalyze/tests/testthat/test-Olink_anova.R:26]: stringr::str_detect

File: OlinkAnalyze/tests/testthat/test-Olink_anova.R [hunk 9]
```diff
@@ -5,6 +5,34 @@ load(refRes_file)
...
+    str_detect(Assay, "Amplification control") ~ "amp_ctrl",
+    str_detect(Assay, "Extension control") ~ "ext_ctrl",
+    TRUE ~ "assay"
+  ))
+
+# Remove control assays from npx_data_format221010 for warning tests
+npx_data_format221010_no_ctrl <- npx_data_format221010 |>
+  dplyr::filter(!str_detect(Assay, "control"))
```
Comment (kathy-nevola) [inline OlinkAnalyze/tests/testthat/test-Olink_anova.R:34]: stringr::str_detect

**Review by kathy-nevola — CHANGES_REQUESTED** [review]
add package names to functions

File: OlinkAnalyze/R/Olink_anova.R (hunk 5 above)
Comment (dtopouza) [reply OlinkAnalyze/R/Olink_anova.R:472]: Done

File: OlinkAnalyze/tests/testthat/test-Olink_anova.R (hunk 6 above)
Comment (dtopouza) [reply OlinkAnalyze/tests/testthat/test-Olink_anova.R:14]: Done

File: OlinkAnalyze/tests/testthat/test-Olink_anova.R (hunk 7 above)
Comment (dtopouza) [reply OlinkAnalyze/tests/testthat/test-Olink_anova.R:25]: Done

File: OlinkAnalyze/tests/testthat/test-Olink_anova.R (hunk 8 above)
Comment (dtopouza) [reply OlinkAnalyze/tests/testthat/test-Olink_anova.R:26]: Done

File: OlinkAnalyze/tests/testthat/test-Olink_anova.R (hunk 9 above)
Comment (dtopouza) [reply OlinkAnalyze/tests/testthat/test-Olink_anova.R:34]: Done

File: OlinkAnalyze/R/Olink_anova.R
[the remark this answers is not in the document]
Comment (dtopouza) [reply OlinkAnalyze/R/Olink_anova.R:1]: Done

**Review by kathy-nevola — APPROVED** [review]

**Pull request merged**